pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DataTypeConversionFunctions.java

```diff
@@ -32,6 +36,51 @@ public class DataTypeConversionFunctions {
   private DataTypeConversionFunctions() {
   }
 
+  @ScalarFunction(nullableParameters = true)
+  public static Object cast(Object value, String targetTypeLiteral) {
+    if (value == null || targetTypeLiteral == null) {
+      return value;
+    }
+    try {
+      Class<?> clazz = value.getClass();
+      boolean isMultiValue = clazz.isArray() & clazz != byte[].class;
+      PinotDataType sourceType = isMultiValue
+          ? PinotDataType.getMultiValueType(clazz)
+          : PinotDataType.getSingleValueType(clazz);
+      PinotDataType targetDataType = convertTypeLiteralToDataType(targetTypeLiteral);
+      if (sourceType == STRING && EnumSet.of(INTEGER, LONG, INTEGER_ARRAY, LONG_ARRAY)
+          .contains(targetDataType)) {
+        if (String.valueOf(value).contains(".")) {
+          // convert integers via double to avoid parse errors
+          if (isMultiValue) {
+            return targetDataType.convert(DOUBLE_ARRAY.convert(value, sourceType), DOUBLE_ARRAY);
+          }
+          return targetDataType.convert(DOUBLE.convert(value, sourceType), DOUBLE);
+        }
+      }
+      return targetDataType.convert(value, sourceType);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Unknown data type: " + targetTypeLiteral);
+    }
+  }
+
+  private static PinotDataType convertTypeLiteralToDataType(String typeLiteral) {
+    String transformed = typeLiteral.toUpperCase();
+    if ("INT".equals(transformed)) {
+      return INTEGER;
+    }
+    if ("INT[]".equals(transformed)) {
+      return INTEGER_ARRAY;
```

Comment:
The target data type should be primitive array

Reply:
I actually hadn't noticed there was also `PRIMITIVE_INTEGER_ARRAY` - let's just drop MV support to make this possible. Not sure how a user would even express an MV literal (though there could be compile time functions which produce MV values).